io/io/test/RIoUring.cxx

```diff
@@ -0,0 +1,51 @@
+#include "ROOT/RIoUring.hxx"
+using RIoUring = ROOT::Internal::RIoUring;
+
+#include "gtest/gtest.h"
+
+TEST(RIoUring, Basics)
+{
+   // successfully construct a ring with queue depth 4
+   RIoUring ring(4);
+}
+
+TEST(RIoUring, IsAvailable)
+{
+   ASSERT_TRUE(RIoUring::IsAvailable());
```

Comment:
Isn't this test going to fail on any platform without iouring? Or does "it compiles" automatically mean "it is available"? But then why do we need `IsAvailable()`?

Reply:
This is sort of an interesting situation where we can really only detect at runtime if `io_uring` works.
We try hard at the cmake level to avoid compiling any `io_uring` stuff if the userspace library `liburing` isn't installed.  
```cmake
if(uring)
  ROOT_ADD_GTEST(RIoUring RIoUring.cxx LIBRARIES RIO uring)
endif()
```

It is possible to imagine a situation where someone installs liburing but doesn't have the right kernel, so it compiles but fails at runtime with `ENOSYS` etc. It's hard to go off kernel versions themselves (maybe there's a backport). 

The test itself was intended to be a warning to the developer that `uring` doesn't work on your machine.

`IsAvailable` is also interesting because right now, all it does it check whether it can open a ring once and caches the result. It is technically possible that this value changes over time (i.e. a number of threads open a ring, exhausting the `memlocked` pool.) I think this different behaviour (checking for each thread) is what's done in the `folly` wrapper. 

https://github.com/facebook/folly/blob/edec54f4d5eac315a278a73d960e97c0de74c503/folly/experimental/io/IoUring.cpp#L193-L203

Reply:
OK but as is, roottest might fail just because uring isn't available in the kernel. I don't think that's a situation we want - I haven't broken my ROOT yet roottest fails, and I cannot fix anything in ROOT to make my roottest work again. Shall we discuss what to do here tomorrow during the I/O meeting?

Reply:
I think by default we should just skip the test if `!IsAvailable()`.  If we want to be very careful, we can set and check an environment variable that indicates that we really expect uring to be there, e.g. `ROOTTEST_HAS_URING`.  On test nodes with a recent enough kernel, we can then fail if this variable is set and `!IsAvailable()`.  To be more safe, we can also revert the default and set `ROOTTEST_NO_URING` where the kernel is too old.

Reply:
I think either approach makes sense and I am OK skipping all the RIoUring tests if it isn't available (and taking out the IsAvailable test entirely).

> I haven't broken my ROOT yet roottest fails

As implemented, people have to opt in to `uring` when they build ROOT. It's their responsibility to have `liburing` installed and have a relatively recent kernel. If they were to then run the `RIoUring` test suite I think it would be strange to have them pass if `uring` isn't working. I feel the way to fix ROOT in this case is to build it again after turning off `uring`.

Reply:
I opened a PR to only test on the dedicated Fedora 32 node here https://github.com/root-project/rootspi/pull/82. This PR then requires changes to make sure we're properly handling the `ROOTTEST_ENABLE_URING` environment variable in `io/io/test/CMakeLists.txt`.